### PR TITLE
[v3] Add padding to `navbar-fixed-*` on modal open

### LINKED ISSUE #27330: [v3] navbar-fixed-* don't get right padding on modal open
When opening a modal and there is a vertical scrollbar on body,
body gets added a pading right of scrollbar width and overflow hidden,  
because without it content will move left and right.  

Same applies for `navbar-fixed-top` and `navbar-fixed-bottom`,  
When body has overflow hidden, `navbar-fixed-*` is moving left and right on modal toggle

- Linux 4.15.0-34-generic #37-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux
- Google Chrome Version 69.0.3497.100 (Official Build) (64-bit)
- https://codepen.io/iamandrewluca/pen/VEZNdY

### PR DESCRIPTION
**Pull request by unrevised6419**
When a modal is opened, body gets some padding if has vertical scrollbars.
Fixed elements on other side don't get this padding, when body has modal-open and it's overflow is hidden.
This will add same amount of padding to fixed elements, if there are vertical scrollbars.
This issue also was addressed in v4, and was fixed and created another one 😄 .
I hope this fix will get into bootstrap-sass

* [x] test added padding
* [x] test removed padding

Fixes: #27330

### CODE REVIEW

File: js/modal.js [hunk 1]
```diff
@@ -269,11 +270,26 @@
...
-    if (this.bodyIsOverflowing) this.$body.css('padding-right', bodyPad + this.scrollbarWidth)
+    var scrollbarWidth = this.scrollbarWidth
+    if (this.bodyIsOverflowing) {
+      this.$body.css('padding-right', bodyPad + scrollbarWidth)
+      $(this.fixedContent).each(function (index, element) {
+        var actualPadding = element.style.paddingRight
+        var calculatedPadding = $(element).css('padding-right')
+        $(element)
+          .data('padding-right', actualPadding)
+          .css('padding-right', parseFloat(calculatedPadding) + scrollbarWidth + `px`)
```
Comment (XhmikosR) [inline js/modal.js]: no need for backticks here.

@Johann-S we should make sure this doesn't happen again in ESLint.

Comment (iamandrewluca) [reply js/modal.js]: done

**Comment by XhmikosR** [comment]
Ah, wait this is for v3. Yeah, v3 isn't in development apart from a new release we'll make for the XSS fixes.

**Comment by iamandrewluca** [comment]
Yeap, I am aware that v3 isn't in development, but it gets bug fixes. I consider this one kind of bug 🙂

**Comment by XhmikosR** [comment]
@Johann-S can you review this?

**Comment by Johann-S** [comment]
We don't include bug fixes, just security fixes, we will ship the last v3 release, so I'm not in favor of merging this PR.

BTW the code looks good, but we include new things, and maybe we'll have to do some more releases which isn't plan at all.

**Comment by iamandrewluca** [comment]
@Johann-S Ok, I get it. Here is a patch that will fix this, if someone wants.

```js
$(function () {
  var fixedContent = '.navbar-fixed-top, .navbar-fixed-bottom';
  var originalSetScrollbar = $.fn.modal.Constructor.prototype.setScrollbar;
  var originalResetScrollbar = $.fn.modal.Constructor.prototype.resetScrollbar;

  $.fn.modal.Constructor.prototype.setScrollbar = function () {
    originalSetScrollbar.call(this);
    var scrollbarWidth = this.scrollbarWidth;
    $(fixedContent).each(function (index, element) {
      var actualPadding = element.style.paddingRight;
      var calculatedPadding = $(element).css('padding-right');
      $(element)
        .data('padding-right', actualPadding)
        .css('padding-right', parseFloat(calculatedPadding) + scrollbarWidth + 'px')
    })
  };


  $.fn.modal.Constructor.prototype.resetScrollbar = function () {
    originalResetScrollbar.call(this);
    $(fixedContent).each(function (index, element) {
      var padding = $(element).data('padding-right');
      $(element).removeData('padding-right');
      element.style.paddingRight = padding ? padding : ''
    })
  };
});
```

**Comment by XhmikosR** [comment]
I think we should make an exception because

1. we have a clear easy to review PR
2. 3.4.0 isn't released yet

**Comment by XhmikosR** [comment]
To be honest I haven't used v3 in any of my projects for like 1 and a half year.

I'm afraid of introducing regressions but if people can test this more, we could include it in 3.4.0.

**Comment by iamandrewluca** [comment]
@XhmikosR added a test, tommorow I'll add one more.

**Comment by XhmikosR** [comment]
Thanks!

**Comment by Johann-S** [comment]
@iamandrewluca do you know if your fix will fix that issue for our v4: https://github.com/twbs/bootstrap/issues/27071 ?

**Comment by iamandrewluca** [comment]
@Johann-S actually my PR is based from v4 PR's. I don't understand why in v4 `.sticky-top` also gets a `margin/padding` on modal open.  
That padding should be added on elements that are based on document viewport, because on modal open scrollbar is hidden and viewport gets bigger.  
`body` element is first one that is based on document viewport, and second one are all fixed positioned elements.
Sticky elements position themself relative to ancestor.

> fixed
The element is removed from the normal document flow, and no space is created for the element in the page layout. It is positioned relative to the initial containing block established by the viewport, except when one of its ancestors has a transform, perspective, or filter property set to something other than none (see the CSS Transforms Spec), in which case that ancestor behaves as the containing block. (Note that there are browser inconsistencies with perspective and filter contributing to containing block formation.) Its final position is determined by the values of top, right, bottom, and left.
https://developer.mozilla.org/en-US/docs/Web/CSS/position

**Comment by Johann-S** [comment]
Hmm ok so according to you on a sticky element we shouldn't apply any margin/padding 🤔 I have to dig on our history to find the reason why we did that, thanks 👍 

BTW if you want to fix that issue on our v4 do not hesitate 😉

**Review by Johann-S — APPROVED** [review]

**Comment by iamandrewluca** [comment]
@Johann-S added one more test. Also I added bootstrap css in `js/test/index.html` some bootstrap logic depends on bootstrap css, for example `Modal.prototype.measureScrollbar` will always return `0` because `.modal-scrollbar-measure` class is used to measure scrollbar width, and has no styles

File: js/tests/index.html [hunk 2]
```diff
@@ -5,6 +5,8 @@
     <title>Bootstrap Plugin Test Suite</title>
     <meta name="viewport" content="width=device-width, initial-scale=1">
 
+    <link rel="stylesheet" href="../../dist/css/bootstrap.css">
```
Comment (Johann-S) [inline js/tests/index.html]: Not a big fan of adding the entire bootstrap css here, I prefer to cherry-pick the CSS rules we need and add it when we need it (and remove them after)

**Comment by iamandrewluca** [comment]
@Johann-S added all bootstrap and removed only for that test 😄 

WARNING: this is a hack.
Adding scrollbar width to other elements on modal open, and removing on close.
```js
var Modal = $.fn.modal.Constructor
var checkScrollbar = Modal.prototype.checkScrollbar
Modal.prototype.checkScrollbar = function () {
  checkScrollbar.call(this)
  this.fixedContent = [this.fixedContent, '.some-fixed-element', '.other-element'].join(', ')
}
```

**Comment by XhmikosR** [comment]
@Johann-S: if you think this is ready feel free to merge.

**Pull request merged**